File: Resources/TOI-Zero_68/A3/008/008.cpp
#include<bits/stdc++.h>
using namespace std;

#define ll long long

int main(){
	ios::sync_with_stdio(0);
	cin.tie(0);
	int n,k;
    cin >> n >> k;
    ll ans=0;
    int a[n];
    for(int i=0;i<n;i++){
    	cin >> a[i];
	}
    deque<pair<int, int> > dq;
    map<int,int> count;
    int sum=0,id=0;
    for(int i=0;i<n;i++){
        if(!count[a[i]]){
        	sum++;
		}
        count[a[i]]++;
        int l=i;
        bool check=false;
        dq.push_back({a[i],i});
        while(sum>=k){
            check=true;
            int x=dq.front().first;
            int idx=dq.front().second;
            l=idx;
            count[x]--;
            if(count[x]==0){
            	sum--;
			}
            dq.pop_front();
        }
        if(l<id || !check){
        	continue;
		}
        ans+=(l-id+1)*(n-1-i+1);
        id=l+1;
    }
    cout << ans;
}

File: Resources/Projects/Archive/2023/Competition/Programming/Contest/TOI-Zero_68/A2/019/TOI-Zero_68/A3/008/008.py
from collections import deque
from collections import defaultdict

n, k = map(int, input().split())
ans = 0
a = list(map(int, input().split()))

dq = deque()
count = defaultdict(int)
sum_unique = 0
id = 0

for i in range(n):
    if count[a[i]] == 0:
        sum_unique += 1
    count[a[i]] += 1
    
    l = i
    check = False
    dq.append((a[i], i))
    
    while sum_unique >= k:
        check = True
        x, idx = dq.popleft()
        l = idx
        count[x] -= 1
        if count[x] == 0:
            sum_unique -= 1
    
    if l < id or not check:
        continue
        
    ans += (l - id + 1) * (n - i)
    id = l + 1

print(ans) 